### ISSUE TITLE
Extremely slow(1/2 hour) mathjax rendering in large notebook

### ISSUE THREAD

**Issue by rlabbe** [ipython/ipython#7952]
I have a very large notebook full of latex (link at bottom). I do not normally work in it directly; I combine 17 or so separate notebooks into one via a python script so that I can generate HTML or PDF for the entire book. 

If I open this notebook by running ipython notebook (jupyter) it takes roughly 1/2 hour for the mathjax to render. I know, mathjax is slow. But it is not that slow. If I open the html versiion of the notebook (generated with 'ipython nbconvert file') it takes maybe 10-20 seconds to render the mathjax. Or, if I load the notebook using nbviewer the mathjax also renders in a suitably short period of time.

In the 1/2 hour of rendering I repeatedly see "Typesetting math: 100%" show up in the lower left, and the notebook scrolls a bit as the layout changes shape. If you scroll through the notebook (painfully slow, but possible) you can see the slow progress down the book as each cell gets rendered. In other words, it is doing work the entire time, it is not just locked up.

This is not my primary way of working with the notebooks, and I know you do not control the mathjax projecs, so I'll certainly understand if this gets marked out of scope or do not fix. But, at least I thought it would be interesting to report. They may be some optimizations that can be done that will help people with more normal sized notebooks. Plus, there are several of us trying to use Jupyter for larger scale book projects like this, so perhaps you'll be interested in this admittedly edge case use.

environment: Windows 7 and xubuntu 14.04 machines. All machines use the Anaconda distribution, and are on IPython 3.0, Python 3.4. My browser is Chrome Version 40.

The notebook is here: 
https://github.com/rlabbe/Kalman-and-Bayesian-Filters-in-Python/blob/master/Kalman_and_Bayesian_Filters_in_Python.ipynb

Fast render via nbviewer here:
http://nbviewer.ipython.org/github/rlabbe/Kalman-and-Bayesian-Filters-in-Python/blob/master/Kalman_and_Bayesian_Filters_in_Python.ipynb

**Comment by rsmith31415** [comment]
I don't think this should be considered an edge case. However, I would suggest to separate your chapters using different ipynb files (as in [Bayesian Methods for Hackers](http://nbviewer.ipython.org/github/CamDavidsonPilon/Probabilistic-Programming-and-Bayesian-Methods-for-Hackers/tree/master/) or [12 steps to Navier-Stokes](http://nbviewer.ipython.org/github/barbagroup/CFDPython/tree/master/lessons/) ). MathJax takes a while to load but also some browsers (Chrome and Firefox) have not been loading big notebooks in a smooth manner.  For example, when I open your nbviewer link in Firefox, it gets unresponsive a couple times before loading correctly.

If you want to create a big PDF, maybe merging your chapters would be a simpler way to do it (pdftk or ghostscript)

**Comment by rlabbe** [comment]
That is what I am doing. This large file is the result of merging many smaller notebooks, one per chapter. I have a Python script that parses the individual notebooks, strips out some cells that I don't want in the final PDF, and creates this master ipynb. But then I would like to open that master ipynb and do things like run calico extensions to number sections, add a TOC, etc., for HTML generation. That is not really workable with the slow load speed. (edit: I think I see what you mean. Generate a PDF for each chapter, then merge the PDF, not the notebooks. I'll try that, thanks).

**Comment by takluyver** [comment]
If it can handle the converted notebook in ~20s, though, it shouldn't be two orders of magnitude slower to render equations in the live notebook. I wonder if it's something like mathjax is getting re-invoked for the whole page after each cell is added to the page.

**Comment by minrk** [comment]
Opening an existing notebook follows the same code path for creating cells and output one by one, as if created in the live notebook. There are significant performance consequences of this for large notebooks. One of the effects is that mathjax rendering is triggered separately for every cell and html or latex output, whereas it's only triggered one time for the whole page in nbconvert.

**Comment by takluyver** [comment]
Hopefully @Carreau's work to achieve a model/view separation will improve this. If that doesn't make it for 4.0, possibly we should consider having an event that fires after loading a notebook is done, and using that to start mathjax running.

**Comment by rsmith31415** [comment]
@rlabbe Yes, that is what I meant. As for the table of contents, the problem is that you want a global table of contents with links working across the whole book, right? I'm not sure if that is possible when you separate the chapters.  
I think it is convenient to think about the benefits of pursuing a very granular global TOC that can be converted to PDF, because it might be tricky to do. Maybe you don't need a very detailed TOC. A reasonable compromise would be to include an additional pynb file with a TOC describing the chapters (but not subsections and sub-subsections and so on) and then convert it to PDF as any other chapter (example: [Numerical Methods](http://nbviewer.ipython.org/github/johntfoster/numerical_methods_book/blob/master/Table%20of%20Contents.ipynb)) and merge everything together to generate a single PDF.

**Comment by rlabbe** [comment]
@rsmith31415

Ah, but that is why we have computers, to automate all that stuff! :) I understand, you are being pragmatic, and certainly the Numerical Methods approach is not a bad thing other than every time you change something you have to go an manually change the TOC (I do have a [TOC](http://nbviewer.ipython.org/github/rlabbe/Kalman-and-Bayesian-Filters-in-Python/blob/master/table_of_contents.ipynb) for nbviewer use). But, for a PDF, what I like is that when I run nbconvert I automatically get a detailed TOC. When viewing it electronically I can click on any line in the toc and it takes me to that page. Furthermore, on the left I get a nice collapsible tree  Those weren't invented in a vacuum - they were invented because for all the advantages you get in the electronic format you lose the physical indexing you get with a physical book. I.e. you can feel and more or less remember that LU decomposition is about 'here', and so on.

And even the the TOC doesn't work *that well* in nbviewer. If I click on John Forester I get this: http://nbviewer.ipython.org/github/johntfoster/  - a big list with no description. If I click on the Numerical Methods repository I get a big listing of chapters, and at the bottom "table of contents". I'd much rather see the info about the book at this level. - Something like a @brief added to the README.md file or something that gets displayed under the repository name. My top level view is even worse, because I have a bunch of sub-directories for support code and such. nbviewer just has no concept of a "book" - a collection of related notebooks. And then I have users reporting problems or asking questions - they can't say "In section 3.4, paragraph 2, ...." because those types of references do not exist in the separate notebook approach. 

I somewhat mischievously suggest the opposite of your suggestion - it is convenient to think of Jupyter supporting longer forms than articles, and some kind of meta-information that allows you to associate several notebooks together as one document.  It's *almost* there.  A nice way to specify sections or notebooks are prefaces or appendices (I hack this with some embedded latex, and a script that parses the .tex file), a footnotes and/or endnotes, spell check, citations, faster performance for larger notebooks, maybe more control over graph layouts (including figure labels like "Fig 2.1 Number of bones given to dogs by weight"), and equation numbering. But hey, everyone is hard at work doing what is clearly more important things like making it language agnostic, so I happily muddle along. Turning it into a proper book publishing system would probably break it for most users, I dunno.

Alan Downey has written about this - he publishes [free books](http://www.greenteapress.com/), but uses the old school latex + separate code files + separate graphs workflow, but has mused about whether IPython Notebook would work for him. I prefer the interactive aspects of Jupyter, and the fact that I don't have to manually ensure that all the code/graph/text is in sync, so I strongly prefer this workflow, but it does break down when you want a true book format in the end. I suspect he can more or less send his tex files to the publisher, whereas I'm not quite sure what the process would be for Jupyter. Certainly the PDF output is not publication quality (bad layout, page breaks orphan things, bad graph sizing, etc). 

I do question whether PDF is even the right way to go for me. I asked another question about Bokeh output to PDF, and it can't be done at this time. I also use IPython widgets, which of course don't work in PDF. Print books are sometimes hard to use because the reader is so separated from the code and charts, and the genius of Jupyter is making it all interactive and responsive. And here I am trying to stuff it in a dead tree format! But, I do know people are opting to read the PDF form instead of the notebook form, and some publishers are printing and selling paper copies of open source books, so there is that as a counter argument. I should probably cut my losses and stick with HTML output ([mine is here](https://github.com/rlabbe/Kalman-and-Bayesian-Filters-in-Python/blob/master/Kalman_and_Bayesian_Filters_in_Python.html), but it has similar problems (no TOC, difficult navigation), so I'd have to script that as well.  Not the end of the world, but I want to work on content, not content delivery. 

Anyway, I don't know if the devs are interested in this discussion; they are focused on other things with broader appeal. I can mostly work around the slow down I reported. I reported it mostly because I thought it might highlight some inefficiency that could be easily fixed, benefiting people with shorter notebooks. And, obviously, I wanted to have this conversation about book production. I hope this doesn't come across as petulant - I love Jupyter, and am just pointing out what problems remain in the book publishing realm for me.

**Comment by takluyver** [comment]
Just to note: we are interested, and using Jupyter to produce longer content, like entire books, is definitely part of the plan. There are a couple of books that have already been written as collections of notebooks - see the books section here: http://nbviewer.ipython.org/

I hope someone who's more familiar with the JS than me will be be able to massively improve performance in situations like this.

**Comment by rsmith31415** [comment]
Sure, the approach I was suggesting requires manual changes although I think it is not too burdensome because you're not likely to add chapters frequently.

I agree that nbviewer doesn't have the correct structure to display a book properly, although I was thinking about the PDF version of your book when I was proposing to write a TOC in an separate ipynb as you already did. 

I would argue that users shouldn't have difficulties reporting issues because all your sections are linkable. However, in this case, the rendering of equations and figures is too slow that hashtag anchors don't work as they should (e.g. [This is a link with a hashtag](http://nbviewer.ipython.org/github/rlabbe/Kalman-and-Bayesian-Filters-in-Python/blob/master/10_Unscented_Kalman_Filter.ipynb#Tracking-Manuevering-Aircraft)). So, yes, that's a problem.

I believe there are markdown extensions to support footnotes and images have an optional title parameter to include a text in a tooltip (`![Alt text](image.jpg "Optional title")`), which already works in Ipython although there are overflow issues. Equation numbering is not working automatically although it is a matter of configuring MathJax to allow AMS-style numbering. I'm not sure if removing automatic numbering was a design decision, though. In any case, better performance when loading large notebooks would be a very welcomed addition.

Actually, I think that you lose significant functionality trying to coerce your lovely interactive book into a PDF. It is a personal preference, but I usually read the notebooks in my browser without even thinking about converting them to PDF and I'd like to see tools including new features that are possible in HTML instead of replicating features of the PDF format.

**Comment by jakobgager** [comment]
@rlabbe just a quick comment... You could generate individual .tex file of all your notebooks using nbconvert with a custom template (to remove the document class, preamble, begin document, etc.).
Later create a master tex document which inputs all these individual files and takes care of the toc. 
This a common way to structure large documents in latex, see e.g. [this wiki](http://en.wikibooks.org/wiki/LaTeX/Modular_Documents)

**Comment by Carreau** [comment]
Closing as this issue is not in IPython itself and if still problematic and relevant should be opened on the right repository.  This will allow to keep the number of opened issue on the IPython repo under control. 

Feel free to keep commenting or reopen if necessary.

Thanks.

**Comment by gsteele13** [comment]
Question: was there another issue reported elsewhere for this? 

I am working on an  automatically generated electronic lab notebook solution for my research group. For documenting 200 files (2 weeks of measurements), we need already 800 cells. Opening this notebook, I end up staring at my browser re-running the mathjax rendering for ages. 

(And to make it worse, the cells don't even contain any formulas...)

Would be mega happy even if there was a way to  for example disable mathjax in the notebook metadata at least...

And if there is another issue based on  this in a  different repo, please let me know, I'll join in there.

Thanks,
Gary

**Comment by Carreau** [comment]
If anything this will likely need to be in https://github.com/jupyter/notebook 
It still is relatively strange that mathjax takes such a long time if there are no formulas. 

Is is as slow in JupyterLaB for example ?

**Comment by gsteele13** [comment]
I just did a test of opening the notebook in both Jupyterlab and in the notebook server, both remotely on my jupyterhub server. 

I timed the  load time with my watch, so not super accurate, but these are the results:

Jupyterlab: 30 seconds
Notebook server: 35 seconds

Ok, it's no half hour at least 😃 but does seem to take a long time. 

For reference, the notebook contains about 600 cells, 400 super-short markdown (heading in one, short sentence in the other), and 200 code cells with one line (with no output). 

I also tried it locally with a notebook server running on my computer, and found a 32 second load time (a bit faster?).

Also, I just tried it  in a local server with mathjax disabled using `--no-mathjax`, and the load time was still 32 seconds. 

So for me at least, the slow rendering is NOT related to mathjax (the 'tooltip' pop-over was just a red herring then...)

It would be nice if the load time of notebooks with loads of cells could be faster. I'll look around to see if there is an open issue for this. 

This is, of course, a non-typical use-case of an autogenerated notebook file. But it maybe there is a quick fix. 

BTW, the code for my logbook generator can be found here in case you are interested:

https://gitlab.tudelft.nl/steelelab/logbook-generator

Cheers,
Gary
